rfcs/global-naming-conventions.md

```diff
@@ -0,0 +1,58 @@
+e.g. parameter names such as `distinct_on` and `order_by`, nested field names such as `stddev_pop`, and type names such as `users_aggregate_fields`. A global setting to use `camelCase` should probably affect these names too.
+
+## Proposal - global `camelCase` option
+
+Add a global option to use `camelCase` for all field names and parameter names generated by Hasura, and to use `PascalCase` for all generated type names.
+We can automatically convert `snake_case` table and schema names to `camelCase` as appropriate.
+
+### Example
+
+For Postgres schema `my_schema` and table `user_accounts` we generate the following root fields:
+
+| Snake case (default)      | With `camelCase` option enabled |
+| --------------------      | ------------------------------- |
+| `my_schema_user_accounts` | `mySchemaUserAccounts` |
+| `my_schema_user_accounts_aggregate` | `mySchemaUserAccountsAggregate` |
+| `my_schema_user_accounts_by_pk` | `mySchemaUserAccountsByPk` |
+| `delete_my_schema_user_accounts` | `deleteMySchemaUserAccounts` |
+| `delete_my_schema_user_accounts_by_pk` | `deleteMySchemaUserAccountsByPk` |
+| `insert_my_schema_user_accounts` | `insertMySchemaUserAccounts` |
+| `insert_my_schema_user_accounts_one` | `insertMySchemaUserAccountsOne` |
+| `update_my_schema_user_accounts` | `updateMySchemaUserAccounts` |
+| `update_my_schema_user_accounts_by_pk` | `updateMySchemaUserAccountsByPk` |
+
+We would also convert all column names to `camelCase`.
+Similarly, all type names would use `PascalCase`, e.g. `my_schema_user_accounts_aggregate_fields` would become `MySchemaUserAccountsAggregateFields`.
+We would also modify other generated names e.g. `distinct_on` -> `distinctOn` and `stddev_pop` -> `stddevPop`.
+
+### Notes
+
+1. If the user specifies a `custom_name` for a table, we should use it as is, without trying to convert to `camelCase`. The user can use this to override the automated case conversion if there are edge cases where it doesn't do what they want.
+However we still need to capitalise the first letter when it's used as the second word of a field name, e.g. `custom_name` of `users` -> field `deleteUsers`.
+1. Users can still override individual root fields names using `custom_root_fields`, e.g. if they really want to use `mySchemaUserAccount` instead of `mySchemaUserAccountsByPk`.
+1. I have only considered Postgres databases so far. Hopefully this proposal should carry over to other backends without problems.
+1. The [Apollo naming conventions](https://www.apollographql.com/docs/apollo-server/schema/schema/#naming-conventions) also mention that enum values should be in `ALL_CAPS`. Enum value names are currently beyond the scope of this RFC.
+
+### Questions
+1. Where should this global option go in the metadata and how should it be set? Or should it be a command line option / environment variable? - Global option in metadata
+1. Do we need an option per source or is a global option better? - global option
```

Comment:
I think having this per source might be better because different databases may have different conventions and the API consumer would like to follow those conventions.

Reply:
Having a per source might give users more fine control at the cost of consistency, especially if the operators also include the naming convention, so that is also something to consider.